alf/algorithms/seditor_algorithm.py

```diff
@@ -0,0 +1,583 @@
+            step_type=inputs.step_type,
+            discount=inputs.discount,
+            actor=actor_info,
+            critic=critic_info,
+            alpha=alpha_loss)
+
+        return AlgStep(output=ap_out.output[0], state=state, info=info)
+
+    def _compute_critics(self,
+                         critic_net,
+                         observation,
+                         action,
+                         replica_min=True):
+        """Compute Q(s,a)"""
+        observation = (observation, action)
+        critics, _ = critic_net(observation)  # [B, replicas * reward_dim]
+        critics = critics.reshape(  # [B, replicas, reward_dim]
+            -1, self._num_critic_replicas, *self._reward_spec.shape)
+        if replica_min:
+            if self.has_multidim_reward():
+                sign = self.reward_weights.sign()
+                critics = (critics * sign).min(dim=1)[0] * sign
+            else:
+                critics = critics.min(dim=1)[0]
+
+        return critics
+
+    def _actor_loss_fn(self, dqda, action):
+        loss = 0.5 * losses.element_wise_squared_loss(
+            (dqda + action).detach(), action)
+        return loss.sum(list(range(1, loss.ndim)))
+
+    def _actor_train_step(self, inputs, ap_out, a_entropy):
+        """Train UM policy according to Eq 7a, to maximize the utility Q.
+        """
+        alpha = self._log_alpha[0].exp().detach()
+
+        critics = self._compute_critics(self._critic_networks,
+                                        inputs.observation, ap_out.output[0])
+        # only maximize the utility Q value
+        q = critics[..., 0].sum()
+
+        dqda = nest_utils.grad(ap_out.a, q)
+        actor_loss = self._actor_loss_fn(dqda, ap_out.a)
+        actor_loss -= alpha * a_entropy
+        return actor_loss
+
+    def _d_actor_train_step(self, inputs, ap_out, da_entropy):
+        """Train SE policy according to Eq 7b, to maximize the constraint Q while
+        minimizing the action editing loss.
+        """
+        alpha = self._log_alpha[1].exp().detach()
+
+        critics = self._compute_critics(self._critic_networks,
+                                        inputs.observation, ap_out.output[1])
+
+        # only maximize the constraint Q value
+        q = critics[..., 1]
+
+        a_l2 = ((ap_out.output[1] - ap_out.output[2])**2).mean(dim=-1)
+
+        if self._regularize_action_diff:
+            # Take mean so that the loss magnitude is invariant to action dimension
+            change_a_loss = a_l2
+        else:
+            with torch.no_grad():
+                a_critics = self._compute_critics(self._critic_networks,
+                                                  inputs.observation,
+                                                  ap_out.output[2])
+            change_a_loss = torch.relu(a_critics[..., 0] - critics[..., 0])
+
+        q2 = torch.stack((-change_a_loss, q), dim=-1)
+        q2 = torch.einsum("bd,d->b", q2, self.reward_weights).sum()
```

Comment:
Is this (before sum) simply a matrix vector product? `torch.matmul(q2, self.reward_weights)` or simply `q2 @ self.reward_weights` should work